### PR TITLE
zfs: add unstable variant

### PR DESCRIPTION
**Pull request by Mic92**
Until now nixos only delivered the latest zfs release. This release is often not
compatible with the latest mainline kernel. Therefor an unstable variant is
added, which might be based on testing releases or git revisions.

fixes #21359

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by edolstra** [comment]
I'm not fond of the `callPackage ./generic.nix` style. IMHO it's cleaner to put all variants in one file and use `callPackages` in `all-packages.nix`. See for example https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/sql/postgresql/default.nix.

**Comment by Mic92** [comment]
done.

File: nixos/modules/tasks/filesystems/zfs.nix [hunk 1]
```diff
@@ -22,12 +22,18 @@ let
 
   kernel = config.boot.kernelPackages;
 
-  splKernelPkg = kernel.spl;
-  zfsKernelPkg = kernel.zfs;
-  zfsUserPkg = pkgs.zfs;
+  packages = if lib.versionAtLeast kernel.kernel.version "4.9" then {
```
Comment (fpletz) [inline nixos/modules/tasks/filesystems/zfs.nix]: I don't think we should install an unstable version of ZFS automatically and without warning if a stable Linux kernel is to be installed.

**Review by fpletz — CHANGES_REQUESTED** [review]
:+1: on all your other changes. Thanks!

File: nixos/modules/tasks/filesystems/zfs.nix (hunk 1 above)
Comment (Mic92) [inline nixos/modules/tasks/filesystems/zfs.nix]: This is indeed questionable.
I will replace it by an assert and a feature flag.

**Pull request merged**